src/java.base/share/classes/java/lang/StableValue.java

```diff
@@ -0,0 +1,675 @@
+ *
+ *     public static Bar bar() {
+ *         return BAR.get();
+ *     }
+ *
+ * }
+ *}
+ * Calling {@code bar()} will create the {@code Bar} singleton if it is not already
+ * created. Upon such a creation, the dependent {@code Foo} will first be created if
+ * the {@code Foo} does not already exist.
+ * <p>
+ * Here is another example where a more complex dependency graph is created in which
+ * integers in the Fibonacci delta series are lazily computed:
+ * {@snippet lang = java:
+ * public final class Fibonacci {
+ *
+ *     private Fibonacci() {}
+ *
+ *     private static final int MAX_SIZE_INT = 46;
+ *
+ *     private static final IntFunction<Integer> FIB =
+ *             StableValue.intFunction(MAX_SIZE_INT, Fibonacci::fib);
+ *
+ *     public static int fib(int n) {
+ *         return n < 2
+ *                 ? n
+ *                 : FIB.apply(n - 1) + FIB.apply(n - 2);
+ *     }
+ *
+ * }
+ *}
+ * Both {@code FIB} and {@code Fibonacci::fib} recurses into each other. Because the
+ * stable int function {@code FIB} caches intermediate results, the initial
+ * computational complexity is reduced from exponential to linear compared to a
+ * traditional non-caching recursive fibonacci method. Once computed, the VM is free to
+ * constant-fold expressions like {@code Fibonacci.fib(10)}.
+ * <p>
+ * The fibonacci example above is a dependency graph with no circular dependencies (i.e.,
+ * it is a dependency tree). If there are circular dependencies in a dependency graph,
+ * a stable value will eventually throw a {@linkplain StackOverflowError} upon referencing
+ * elements in a circularity.
+ *
+ * <h2 id="thread-safety">Thread Safety</h2>
+ * The content of a stable value is guaranteed to be set at most once. If competing
+ * threads are racing to set a stable value, only one update succeeds, while other updates
+ * are blocked until the stable value becomes set.
+ * <p>
+ * The at-most-once write operation on a stable value that succeeds
+ * (e.g. {@linkplain #trySet(Object) trySet()})
+ * {@linkplain java.util.concurrent##MemoryVisibility <em>happens-before</em>}
+ * any subsequent read operation (e.g. {@linkplain #orElseThrow()}).
+ * <p>
+ * The method {@linkplain StableValue#orElseSet(Supplier) orElseSet()} guarantees that
+ * the provided {@linkplain Supplier} is invoked successfully at most once even under
+ * race. Since stable functions and stable collections are built on top of
+ * {@linkplain StableValue#orElseSet(Supplier) orElseSet()} they too are
+ * thread safe and guarantee at-most-once-per-input invocation.
+ *
+ * <h2 id="miscellaneous">Miscellaneous</h2>
+ * Except for a StableValue's content itself, an {@linkplain #orElse(Object) orElse(other)}
+ * parameter, and an {@linkplain #equals(Object) equals(obj)} parameter; all method
+ * parameters must be <em>non-null</em> or a {@link NullPointerException} will be thrown.
+ * <p>
+ * Stable functions and collections are not {@link Serializable} as this would require
```

Comment:
Not sure this belongs here. Perhaps the comment on these functions not being serializable should be on their factories. And the point on security vulnerability seems specific and vague at the same time -- better remove it.